//protoc --cpp_out dllexport_decl=JDE_MARKETS_EXPORT:. requests.proto
//npx pbjs -r request_root -t static-module -w es6 -o src/app/proto/requests.js requests.proto & npx pbts -o src/app/proto/requests.d.ts src/app/proto/requests.js
//npx pbjs -r request_root -t static-module -w es6 -o requests.js /home/duffyj/code/jde/MarketLibrary/source/types/proto/requests.proto & npx pbts -o requests.d.ts requests.js

syntax = "proto3";
package Jde.Markets.Proto.Requests;
import "ib.proto";

message GenericRequests
{
	ERequests type=1;
	repeated int32 ids=2;
}

message RequestAccountUpdates
{
	bool subscribe=1;
	string account_number=2;
}

message RequestAccountUpdatesMulti
{
	int32 id=1;
	string account_number=2;
	string model_code=3;
	bool ledger_and_nlv=4;
}

enum ETickList
{
	TickListNone=0;
	OptionVolume=100;
	OptionOpenInterest=101;
	AverageOptVolume=105;
	Impvolat=106;
	Climpvlt=107;
	Bondanalyticdata=125;
	MiscStats=165;
	CScreen=166;
	CreditmanMarkPrice=221;
	Auction=225;
	PlPrice=232;
	RTVolume=233;
	Inventory=236;
	Fundamentals=258;
	Ivclose=291;
	WideNews=292;
	TradeCount=293;
	TradeRate=294;
	VolumeRate=295;
	LastRTHTrade=318;
	ParticipationMonitor=370;
	RTTrdVolume=375;
	CttTickTag=377;
	IBRate=381;
	RfqTickRespTag=384;
	DMM=387;
	IssuerFundamentals=388;
	IBWarrantImpVolCompeteTick=391;
	IndexCapabilities=405;
	FuturesMargins=407;
	rthistvol=411;
	MonitorTickTag=439;
	RTCLOSE=459;
	BondFactorMultiplier=460;
	FeeandRebateRate=499;
	midptiv=506;
	hvolrt10perUnderlying=511;
	hvolrt30perUnderlying=512;
	hvolrt50perUnderlying=513;
	hvolrt75perUnderlying=514;
	hvolrt100perUnderlying=515;
	hvolrt150perUnderlying=516;
	hvolrt200perUnderlying=517;
	fzmidptiv=521;
	vsiv=545;
	EtfNavBidAsknavbidask=576;
	EtfNavLastnavlast=577;
	EtfNavClosenavclose=578;
	AverageOpeningVol=584;
	AverageClosingVol=585;
	PlPriceDelayed=587;
	FuturesOpenInterest=588;
	EMAN=608;
	EtfNavMischightLow=614;
	CreditmanSlowMarkPrice=619;
	EtfFrozenNavLastfznavlast=623;
	MonetaryClosePrice=645;///428
	Avgv1Min=658;
}

message RequestHistoricalData
{
	int32 id=1;
	Contract contract=2;
	int32 date=3;
	uint32 days=4;
	BarSize bar_size=5;
	Display display=6;
	bool use_rth=7;
	bool keep_up_to_date=8;
}
message RequestMrkDataSmart
{
	int32 id=1;
	int32 contract_id=2;
	repeated ETickList tick_list=3;
	bool snapshot=4;
}

message RequestContractDetails
{
	int32 id=1;
	repeated Contract contracts=2;
}

message RequestOptions
{
	uint32 id=1;
	uint32 contract_id=2;
	boolean is_call=3;
	uint32 date=4;
}

message FlexExecutions
{
	int32 id=1;
	string account_number=2;
	uint32 start=3;
	uint32 end=4;
}


message PlaceOrder
{
	uint32 id=1;
	Contract contract=2;
	Order order=3;
	double stop=4;
	double stop_limit=5;
}

message RequestUnion
{
	oneof Value
	{
		RequestAccountUpdates account_updates = 1;
		RequestAccountUpdatesMulti account_updates_multi = 2;
		RequestContractDetails contract_details = 3;
		GenericRequests generic_requests=4;
		RequestMrkDataSmart MrkDataSmart = 5;
		RequestOptions options = 7;
		RequestHistoricalData HistoricalData = 8;
		FlexExecutions flex_executions=9;
		PlaceOrder place_order=10;
		//Request Historical data like normal...
	}
}

message RequestTransmission
{
	repeated RequestUnion Messages=10;
}

enum Display
{
	Trades=0;
	Midpoint=1;
	Bid=2;
	Ask=3;
	BidAsk=4;
	HistoricalVolatility=5;
	OptionImpliedVolatility=6;
	FeeRate=7;
	RebateRate=8;
}

enum BarSize
{
	None=0;
	Second=1;
	Second5=2;
	Second15=3;
	Second30=4;
	Minute=5;
	Minute2=6;
	Minute3=16;
	Minute5=7;
	Minute15=8;
	Minute30=9;
	Hour=10;
	Day=11;
	Week=12;
	Month=13;
	Month3=14;
	Year=15;
}

enum ERequests
{
	Ping									= 0;

	MarketData             			= 1;
	CancelMarketData              = 2;
	PlaceOrder_                  	= 3;
	CancelOrder                  	= 4;
	RequestOpenOrders             = 5;
	REQ_ACCT_DATA                 = 6;
	REQ_EXECUTIONS                = 7;
	RequestIds                    = 8;
	REQ_CONTRACT_DATA             = 9;
	REQ_MKT_DEPTH                 = 10;
	CANCEL_MKT_DEPTH              = 11;
	REQ_NEWS_BULLETINS            = 12;
	CANCEL_NEWS_BULLETINS         = 13;
	SET_SERVER_LOGLEVEL           = 14;
	REQ_AUTO_OPEN_ORDERS          = 15;
	RequestAllOpenOrders          = 16;
	ManagedAccounts        			= 17;
	REQ_FA                        = 18;
	REPLACE_FA                    = 19;
	REQ_HISTORICAL_DATA           = 20;
	EXERCISE_OPTIONS              = 21;
	REQ_SCANNER_SUBSCRIPTION      = 22;
	CANCEL_SCANNER_SUBSCRIPTION   = 23;
	REQ_SCANNER_PARAMETERS        = 24;
	CANCEL_HISTORICAL_DATA        = 25;
	CurrentTime                   = 49;
	RequestRealTimeBars           = 50;
	CANCEL_REAL_TIME_BARS         = 51;
	REQ_FUNDAMENTAL_DATA          = 52;
	CANCEL_FUNDAMENTAL_DATA       = 53;
	REQ_CALC_IMPLIED_VOLAT        = 54;
	REQ_CALC_OPTION_PRICE         = 55;
	CANCEL_CALC_IMPLIED_VOLAT     = 56;
	CANCEL_CALC_OPTION_PRICE      = 57;
	REQ_GLOBAL_CANCEL             = 58;
	REQ_MARKET_DATA_TYPE          = 59;
	Positions              			= 61;
	REQ_ACCOUNT_SUMMARY           = 62;
	CANCEL_ACCOUNT_SUMMARY        = 63;
	CANCEL_POSITIONS              = 64;
	VERIFY_REQUEST                = 65;
	VERIFY_MESSAGE                = 66;
	QUERY_DISPLAY_GROUPS          = 67;
	SUBSCRIBE_TO_GROUP_EVENTS     = 68;
	UPDATE_DISPLAY_GROUP          = 69;
	UNSUBSCRIBE_FROM_GROUP_EVENTS = 70;
	StartApi                      = 71;
	VERIFY_AND_AUTH_REQUEST       = 72;
	VERIFY_AND_AUTH_MESSAGE       = 73;
	REQ_POSITIONS_MULTI           = 74;
	CANCEL_POSITIONS_MULTI        = 75;
	RequestAccountUpdatesMulti_   = 76;
	CANCEL_ACCOUNT_UPDATES_MULTI  = 77;
	REQ_SEC_DEF_OPT_PARAMS        = 78;
	REQ_SOFT_DOLLAR_TIERS         = 79;
	REQ_FAMILY_CODES              = 80;
	REQ_MATCHING_SYMBOLS          = 81;
	REQ_MKT_DEPTH_EXCHANGES       = 82;
	REQ_SMART_COMPONENTS          = 83;
	REQ_NEWS_ARTICLE              = 84;
	REQ_NEWS_PROVIDERS            = 85;
	REQ_HISTORICAL_NEWS           = 86;
	REQ_HEAD_TIMESTAMP            = 87;
	REQ_HISTOGRAM_DATA            = 88;
	CANCEL_HISTOGRAM_DATA         = 89;
	CANCEL_HEAD_TIMESTAMP         = 90;
	REQ_MARKET_RULE               = 91;
	REQ_PNL                       = 92;
	CANCEL_PNL                    = 93;
	REQ_PNL_SINGLE                = 94;
	CANCEL_PNL_SINGLE             = 95;
	REQ_HISTORICAL_TICKS          = 96;
	REQ_TICK_BY_TICK_DATA         = 97;
	CANCEL_TICK_BY_TICK_DATA      = 98;
}

